docs/source/reference/envs.rst

```diff
@@ -169,7 +186,37 @@ one can simply call:
         >>> print(a)
         9.81
 
-It is also possible to reset some but not all of the environments:
+TorchRL uses a private ``"_reset"`` key to indicate to the environment which
+sub-environments should be reset.
+This allows to reset some but not all of the environments.
+Keep in mind that this is a private key, and it should only be used when coding
+specific environment features that are internal facing.
+The way an environment deals with the ``"_reset"`` keys is proper to it and
+designing an environment that behaves according to ``"_reset"`` inputs is the
+developer's responsibility. The following assumptions are made:
+
+- Each ``"_reset"`` is paired with a ``"done"`` (and/or ``"terminated"``,
+  ``"truncated"`` etc.). This means that the following structure is not
+  allowed: ``TensorDict({"done": done, "nested": {"_reset": reset}}, [])``.
+- A reset at one level precludes the presence of a ``"_reset"`` at lower
+  levels below it. For instance, this is not allowed:
+  ``TensorDict({"_reset": reset_main, "nested": {"_reset": reset_leaf}}, [])``
+- When calling :meth:`env.reset(tensordict)` with a partial ``"_reset"`` entry
+  that will reset some but not all the done sub-environments, the input data
+  should contain the data of the sub-environemtns that are __not__ being reset.
+  The reason for this constrain is that the output of the ``env._reset(data)``
+  can only be predicted for the entries that are reset. For the others, we
+  cannot know in advance if they will be meaningful or not. For instance,
+  one could perfectly just pad the values of the non-reset environments, in
+  which case the data will be meaningless and should be discarded.
+  :meth:`~torchrl.envs.EnvBase.reset` will merge the input tensordict with
+  the one resulting from reset and pad missing values with 0 (except for
+  ``done`` signals).
```

Comment:
I don't see why if we cannot predict if they will be meaningful we automatically pad with 0s

It seems like the options are:
- pad with the values from `env._reset()` which could be meaningful or not
- pad with 0 which is certainly not meaningful

This is a very simple choice for me. What am I missing?
My suggestion is to remove any 0 padding and simlply use data from `env._reset()` whenever we would have padded

Reply:
That does not work unfortunately. 
We have no idea about what is going on within `_reset` in general, and per se the only thing we want to ask is that `"_reset"` points to what is being changed. For the rest we assume that the data is unchanged compared to the last call to `step`. We have access to that data so we can copy it. With stateless environment, not padding isn't really an option I think.

Reply:
For instance, [here](https://github.com/pytorch/rl/blob/38dfc21758f57683071fc29114c338d12f4b8d31/torchrl/envs/batched_envs.py#L624-L640) we assume that the non-reset data is identical to the content of the buffer, but that's an implicit assumption and a hard one to enforce/test. 
It also introduces some unwanted overhead since we will be calling where in the collector (not anymore with this PR since we do the job just once).
What we're suggesting with this PR is to do the call of `torch.where` in a consistent way, just once after calling `reset`

Reply:
Another thought:
If we say "anything returned along the non-reset axis will be considered as valid data", we're not telling the truth since the collector currently discards that data in favour of the latest "next" data using `torch.where`. So either we say the opposite ("anything returned along the non-reset axis will be considered as **in**valid data") or we say "sometimes one, sometimes the other"

Reply:
> "anything returned along the non-reset axis will be considered as valid data"

I am not saying this.

I am saying that in the extreme case where `env.reset()` would pad, we fallback to it.

See my comments in the other thread.

Reply:
again, i understand that, but that's the same as saying that we consider that data as "valid".
Let me give you an example: in this PR, i discovered many places where we do a partial reset and check that the non-reset data is the same as the data in the previous next. This is assuming that the env is stateful, and as such it's a bad decision (a bad thing to test). We should not assume that at any place in the code. Padding with 0s makes sure that if we assume that, we're wrong, and not "sometimes right", which is actually the key to prevent silent bugs.

Reply:
If reading data from non-reset entries is not supported, we should lock that feature. 
For instance: if you call `.backward()` on a value, you don't expect tensors that are out of the graph to have non-zero gradients assigned to them. If you are using a MaskedTensor, you don't expect any operation to produce meaningful values for masked entries. This is the same logic here, we lock this logic by padding (or better: masking) with 0s anything that is meaningless. it's a pretty straightforward contract I think

Reply:
Here's another example:
@albertbou92 recently uncovered a bug where the following code fails
```python
torch.manual_seed(0)
env = ParallelEnv(2, lambda: TransformedEnv(GymEnv("CartPole-v1"), CatFrames(in_keys=["observation"], out_keys=["observation_cat"], N=4, dim=-1)))
env.set_seed(0)
r = env.rollout(1000, break_when_any_done=False)

torch.manual_seed(0)
env = TransformedEnv(ParallelEnv(2, lambda: GymEnv("CartPole-v1")), CatFrames(in_keys=["observation"], out_keys=["observation_cat"], N=4, dim=-1))
env.set_seed(0)
r2 = env.rollout(1000, break_when_any_done=False)

torch.testing.assert_close(r["observation_cat"], r2["observation_cat"])
```

This is caused by `CatFrames` having a buffer stored as `nn.Module._buffer` instance where we write data as it flows. But when we call a partial `reset`, we also call `transform._call` which applies the transform to all dimensions (not only the part of the data that is being reset). That means that we corrupt data for envs not being reset.
The solution I'm contemplating is to carry the buffer in the tensordict and not in the transform. 
That way, when we output the result of `CatFrames._reset`, we get some corrupted data that we can override with the data from the prev step. 
I think that presenting non-masked data that is corrupted is wrong n this case. It will make it harder to test that the class its doing its job (with the solution I propose we can just check that maskes data is 0 and that the rest respects the convention of the CatFrame transform).
It's a very specific example but I think there are many others where the call to `transform._call` within the transform class `_reset` method will corrupt data in a way that we want to get rid of

Reply:
> in this PR, i discovered many places where we do a partial reset and check that the non-reset data is the same as the data in the previous next. This is assuming that the env is stateful, and as such it's a bad decision (a bad thing to test)

This to me is the most important test regarding resets.
Sure, if we are using this in non-stateful enviornments, this test has to be adapted, but for stateful ones to me this is a core thing to check.

Reply:
See my comment below: I think this _clarifies_ these tests as we're not assuming anything regarding the behaviour of non-reset entries.